### PR TITLE
Fix location of Integrated areas derivatives in the PCG solver

### LINKED ISSUE #2708: Potential bug introduced either in immersed boundaries or at PGC (or somewhere else)?
Something happened from v0.76.8 to v0.77.1. In particular, I discovered that a horizontally regular domain with an immersed boundary and a hydrostatic free-surface model with a PGC solver produces inconsistent results!

On Julia v1.6 with **Oceananigans v0.76.8** I run this [script](https://github.com/navidcy/multigrid_free_surface_solvers_benchmarks/blob/ncc/Oceananigans-v0.76.8/implicit_surface_solvers_benchmark.jl) and get:

```Julia
julia> include("implicit_surface_solvers_benchmark.jl")
[ Info: Oceananigans will use 6 threads
┌ Warning: defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended
└ @ Oceananigans.Advection ~/.julia/packages/Oceananigans/k6G5J/src/Advection/weno_fifth_order.jl:187
┌ Warning: defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended
└ @ Oceananigans.Advection ~/.julia/packages/Oceananigans/k6G5J/src/Advection/weno_fifth_order.jl:187
[ Info: Running simulation:
[ Info: Initializing simulation...
┌ Info: Iter: 0, time: 0 seconds, wall time: 7.325 seconds, max|w|: 0.000e+00, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 0
[ Info:     ... simulation initialization complete (1.556 seconds)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (1.313 minutes).
┌ Info: Iter: 10, time: 1.667 hours, wall time: 1.367 minutes, max|w|: 3.470e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 73
┌ Info: Iter: 20, time: 3.333 hours, wall time: 337.826 ms, max|w|: 5.644e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 75
┌ Info: Iter: 30, time: 5 hours, wall time: 315.972 ms, max|w|: 5.883e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 79
┌ Info: Iter: 40, time: 6.667 hours, wall time: 337.502 ms, max|w|: 4.817e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 76
┌ Info: Iter: 50, time: 8.333 hours, wall time: 334.497 ms, max|w|: 3.370e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 79
┌ Info: Iter: 60, time: 10 hours, wall time: 332.201 ms, max|w|: 3.190e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 78
┌ Info: Iter: 70, time: 11.667 hours, wall time: 343.445 ms, max|w|: 4.322e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 79
┌ Info: Iter: 80, time: 13.333 hours, wall time: 339.778 ms, max|w|: 5.128e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 90, time: 15 hours, wall time: 329.197 ms, max|w|: 5.246e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 77
┌ Info: Iter: 100, time: 16.667 hours, wall time: 340.226 ms, max|w|: 4.030e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 79
┌ Info: Iter: 110, time: 18.333 hours, wall time: 337.669 ms, max|w|: 3.692e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 80
┌ Info: Iter: 120, time: 20 hours, wall time: 345.395 ms, max|w|: 4.089e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 80
┌ Info: Iter: 130, time: 21.667 hours, wall time: 350.350 ms, max|w|: 4.690e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 140, time: 23.333 hours, wall time: 334.162 ms, max|w|: 4.542e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 150, time: 1.042 days, wall time: 347.559 ms, max|w|: 3.833e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 160, time: 1.111 days, wall time: 346.541 ms, max|w|: 3.322e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 170, time: 1.181 days, wall time: 352.355 ms, max|w|: 3.220e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 180, time: 1.250 days, wall time: 350.246 ms, max|w|: 3.838e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 190, time: 1.319 days, wall time: 352.536 ms, max|w|: 3.623e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 200, time: 1.389 days, wall time: 383.801 ms, max|w|: 3.135e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 210, time: 1.458 days, wall time: 332.942 ms, max|w|: 3.084e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 220, time: 1.528 days, wall time: 351.647 ms, max|w|: 3.197e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 230, time: 1.597 days, wall time: 348.245 ms, max|w|: 4.172e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 240, time: 1.667 days, wall time: 352.064 ms, max|w|: 4.340e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 250, time: 1.736 days, wall time: 353.158 ms, max|w|: 3.628e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 260, time: 1.806 days, wall time: 352.907 ms, max|w|: 3.765e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 270, time: 1.875 days, wall time: 340.669 ms, max|w|: 3.594e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 280, time: 1.944 days, wall time: 350.226 ms, max|w|: 4.015e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
[ Info: Simulation is stopping. Model time 2 days has hit or exceeded simulation stop time 2 days.
```

On Julia v1.6 with **Oceananigans v0.77.1** I run this [script](https://github.com/navidcy/multigrid_free_surface_solvers_benchmarks/blob/ncc/Oceananigans-v0.77.1/implicit_surface_solvers_benchmark.jl) (which is slightly modified to account for changes between v0.76.8 and v0.77.0; namely `WENO5` -> `WENO`) and get:

```julia
julia> include("implicit_surface_solvers_benchmark.jl")
[ Info: Running simulation:
[ Info: Initializing simulation...
┌ Info: Iter: 0, time: 0 seconds, wall time: 7.947 seconds, max|w|: 0.000e+00, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 0
[ Info:     ... simulation initialization complete (1.652 seconds)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (1.788 minutes).
┌ Info: Iter: 10, time: 1.667 hours, wall time: 1.883 minutes, max|w|:    NaN, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 4096
┌ Info: Iter: 20, time: 3.333 hours, wall time: 12.369 seconds, max|w|:    NaN, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 4096
┌ Info: Iter: 30, time: 5 hours, wall time: 12.496 seconds, max|w|:    NaN, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 4096
┌ Info: Iter: 40, time: 6.667 hours, wall time: 12.428 seconds, max|w|:    NaN, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 4096
```

As a last attempt I tried changing the halos from 4 to 5 on **Oceananigans v0.77.1**. Didn't change anything.
(cc @elise-palethorpe, @sandreza)

### PR DESCRIPTION
**Pull request by simone-silvestri**
closes #2708

### CODE REVIEW

File: src/Models/HydrostaticFreeSurfaceModels/pcg_implicit_free_surface_solver.jl [hunk 1]
```diff
@@ -44,8 +44,8 @@ step `Δt`, gravitational acceleration `g`, and free surface at time-step `n` `
 """
 function PCGImplicitFreeSurfaceSolver(grid::AbstractGrid, settings, gravitational_acceleration=nothing)
     # Initialize vertically integrated lateral face areas
-    ∫ᶻ_Axᶠᶜᶜ = Field{Face, Center, Nothing}(with_halo((3, 3, 1), grid))
-    ∫ᶻ_Ayᶜᶠᶜ = Field{Center, Face, Nothing}(with_halo((3, 3, 1), grid))
+    ∫ᶻ_Axᶠᶜᶜ = Field{Face, Center, Nothing}(grid)
+    ∫ᶻ_Ayᶜᶠᶜ = Field{Center, Face, Nothing}(grid)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/pcg_implicit_free_surface_solver.jl:48]: ?

**Comment by glwagner** [comment]
We should add a test to prevent this (and bugs like this) from entering in the future.

File: src/Models/HydrostaticFreeSurfaceModels/pcg_implicit_free_surface_solver.jl (hunk 1 above)
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/pcg_implicit_free_surface_solver.jl]: a try to make it work before finding the real issue

**Comment by navidcy** [comment]
shall I test this?

**Comment by navidcy** [comment]
OK, with this running this [script](https://github.com/navidcy/multigrid_free_surface_solvers_benchmarks/blob/ncc/Oceananigans-v0.77.1/implicit_surface_solvers_benchmark.jl) I get:

```Julia
julia> include("implicit_surface_solvers_benchmark.jl")
[ Info: Oceananigans will use 6 threads
[ Info: Running simulation:
[ Info: Initializing simulation...
┌ Info: Iter: 0, time: 0 seconds, wall time: 7.674 seconds, max|w|: 0.000e+00, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 0
[ Info:     ... simulation initialization complete (1.722 seconds)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (1.799 minutes).
┌ Info: Iter: 10, time: 1.667 hours, wall time: 1.853 minutes, max|w|: 3.674e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 73
┌ Info: Iter: 20, time: 3.333 hours, wall time: 399.705 ms, max|w|: 5.917e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 75
┌ Info: Iter: 30, time: 5 hours, wall time: 369.868 ms, max|w|: 6.080e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 79
┌ Info: Iter: 40, time: 6.667 hours, wall time: 389.075 ms, max|w|: 4.834e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 76
┌ Info: Iter: 50, time: 8.333 hours, wall time: 422.587 ms, max|w|: 2.943e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 79
┌ Info: Iter: 60, time: 10 hours, wall time: 381.054 ms, max|w|: 2.652e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 78
┌ Info: Iter: 70, time: 11.667 hours, wall time: 413.078 ms, max|w|: 4.074e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 79
┌ Info: Iter: 80, time: 13.333 hours, wall time: 391.253 ms, max|w|: 5.141e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 90, time: 15 hours, wall time: 386.768 ms, max|w|: 4.920e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 77
┌ Info: Iter: 100, time: 16.667 hours, wall time: 411.517 ms, max|w|: 3.374e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 79
┌ Info: Iter: 110, time: 18.333 hours, wall time: 386.204 ms, max|w|: 2.843e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 80
┌ Info: Iter: 120, time: 20 hours, wall time: 419.526 ms, max|w|: 3.045e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 80
┌ Info: Iter: 130, time: 21.667 hours, wall time: 395.954 ms, max|w|: 3.905e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 140, time: 23.333 hours, wall time: 389.321 ms, max|w|: 4.118e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 150, time: 1.042 days, wall time: 425.186 ms, max|w|: 3.476e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 160, time: 1.111 days, wall time: 395.464 ms, max|w|: 2.492e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 170, time: 1.181 days, wall time: 410.511 ms, max|w|: 2.896e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 180, time: 1.250 days, wall time: 396.903 ms, max|w|: 3.124e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 190, time: 1.319 days, wall time: 396.648 ms, max|w|: 3.279e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 200, time: 1.389 days, wall time: 422.681 ms, max|w|: 3.090e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 210, time: 1.458 days, wall time: 387.975 ms, max|w|: 2.255e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 220, time: 1.528 days, wall time: 411.532 ms, max|w|: 2.715e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 81
┌ Info: Iter: 230, time: 1.597 days, wall time: 391.716 ms, max|w|: 3.341e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 240, time: 1.667 days, wall time: 393.949 ms, max|w|: 3.202e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 250, time: 1.736 days, wall time: 455.565 ms, max|w|: 3.156e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 83
┌ Info: Iter: 260, time: 1.806 days, wall time: 397.246 ms, max|w|: 2.778e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 270, time: 1.875 days, wall time: 417.887 ms, max|w|: 2.195e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
┌ Info: Iter: 280, time: 1.944 days, wall time: 399.195 ms, max|w|: 3.006e-03, m s⁻¹, next Δt: 10 minutes
└ solver iterations: 82
[ Info: Simulation is stopping. Model time 2 days has hit or exceeded simulation stop time 2 days.
```

I doubt it's identical to the v0.76.8 version but it's not blowing up!

**Review by navidcy — APPROVED** [review]

**Comment by navidcy** [comment]
merge?

**Comment by glwagner** [comment]
> I agree; I also think this shows the need for a more structured definition of a `Nothing` location
> 
> We should split it into two in order to have two different `ReducedFields`:
> 
> * a _truly_ 2D field (for reductions)
> * a 2.5D field with an index associated with the reduced dimension (for free surface or similar fields)

We introduced `Field.indices` to solve this problem. The 2.5D field would have `indices = (:, :, grid.Nz)`. Does this work?

**Comment by glwagner** [comment]
Did we add a test?

**Comment by navidcy** [comment]
Nope. I'll open a new PR for that!

**Pull request merged**